ACPs/194-streaming-asynchronous-execution/README.md

```diff
@@ -118,49 +125,123 @@ ACP-103 introduced the following variables for calculating the gas price:
+### Handling gas target changes
 
-To achieve this, we first define the _executed_ and _settled_ states for blocks.
-A block is initially _accepted_ by consensus, enqueuing its tranche of transactions.
-Once the last transaction in the tranche has been executed, the block itself is considered to be executed.
-When the same transaction is settled then so too is the block considered settled.
+When a block is produced which modifies $T$, both the consensus thread and the execution thread will update to the modified $T$ after their own handling of the block.
 
-#### Named blocks
+For example, restrictions of the queue size MUST be calculated based on the parent block's $T$.
 
-The Ethereum Mainnet APIs allow for retrieving blocks by named parameters that the API server resolves based on their consensus mechanism.
-Other than the _earliest_ (genesis) named block, which MUST be interpreted in the same manner, all other named blocks are mapped to SAE in terms of the execution status of all block transactions and MUST be interpreted as follows:
+Similarly, the time spent executing a block MUST be calculated based on the parent block's $T$.
 
- * _pending_: the most recently _accepted_ block;
- * _latest_: the block that was most recently _executed_;
- * _safe_ and _finalized_: the block that was most recently _settled_.
+### Block settlement
+
+For a _proposed_ block which includes timestamp $t_b$, all ancestors whose execution timestamp $t_e$ is $t_e + \tau \leq t_b$ are considered settled.
+
+The _proposed_ block must include the `stateRoot` produced by the execution of the most recently settled block.
+
+For any _newly_ settled blocks, the _proposed_ block must include all execution artifacts:
+- `receiptsRoot`
+- `logsBloom`
+- `gasUsed`
 
 > [!NOTE]
-> The finality guarantees of Snowman consensus remove any distinction between _safe_ and _finalized_. 
-> Furthermore, the _latest_ block is not at risk of re-org, only of a negligible risk of data corruption local to the API node.
+> If the block executor has fallen behind, the node may not be able to determine precisely which ancestors should be considered settled. If this occurs, validators MUST allow the block executor to catch up prior to deciding the block's validity.
+
+### Block validity and building
+
+After determining which blocks to settle, all remaining ancestors of the new block must be inspected to determine the worst-case bounds on $x$ and account balances. Account nonces are able to be known immediately.
+
+The worst-case bound on $x$ can be calculated by following the block executor update rules using $g_L$ rather than $g_C$.
+
+The worst-case bound on account balances can be calculated by charging the worst-case gas cost to the sender of a transaction along with deducting the value of the transaction from the sender's account balance.
+
+The `baseFeePerGas` field must be populated to the gas price based on the worst-case bound on $x$ at the start of block execution.
```

Comment:
```suggestion
The `baseFeePerGas` field MUST be populated with the gas price based on the worst-case bound on $x$ at the start of block execution.
```

Interesting. This could be problematic when later returning an executed block from the API as the actual charge will differ. I think we should consider adding a new header field for `estimatedBaseFeePerGas` (or similar) and leaving `baseFeePerGas` as zero for consensus, only populating it when returned by APIs.

I know that adding an extra field is a decision that shouldn't be taken lightly, but it's definitely worth considering the tradeoff. From an engineering perspective it's now easy with libevm's RLP-override capability.

Reply:
I felt like we had two options here:
1. `baseFeePerGas` is the gasPrice of the most recently settled block
2. `baseFeePerGas` is the worst-case gasPrice bound of the current block

The second had the following nice properties:
- All transactions in the block will have a baseFee >= the baseFee in the header
- Some (many) wallets use the latest block's baseFee as a fee estimator

If we wanted to override the API results, I think we can consider that in the appendix... This might be a good question to bring up on the community call.

Reply:
> All transactions in the block will have a baseFee >= the baseFee in the header

Do you mean `<=`?

The first has an undesirable property of not applying to all the transactions being settled.

> If we wanted to override the API results, I think we can consider that in the appendix... This might be a good question to bring up on the community call.

Agreed so leaving unresolved to remind us.

Reply:
Do you mean <=?

The will be _charged_ a baseFee <= the baseFee in the block. But the transactions are enforced to specify internally a baseFee >= the baseFee in the block.